MetaMask/metamask-extension
# Allow adding networks with the same chainId as a preloaded/default network via `wallet_AddEthereumChain` API

* Fixes #16712

Currently within the add-custom-network form we allow users to add networks with the same chainId as one of the default networks we preload - including Ethereum Mainnet. As part of our efforts to enhance users' privacy we want to allow users to add alternative Ethereum Mainnet RPC providers via the `wallet_AddEthereumChain` API. 

But because this is a potentially sensitive action - dapp's might propose untrustworthy or malicious RPC providers - we want to add additional warnings and friction to ensure that they have more opportunities to inform themselves of the risks associated with alternative (non-infura) mainnet RPC providers. So this PR adds an additional warning modal that pops when users click `approve` on the initial `Allow this site to add a network?` notification. This modal provides further information and warnings about the potential risks associated with this action:

<img width="372" alt="Screenshot 2022-12-01 at 11 02 21 AM" src="https://user-images.githubusercontent.com/34557516/205114431-790bdd08-2e52-498a-9326-5bbd33dbb1d8.png">

TODO:
- [ ] ["Templatize" the warning modal](https://github.com/MetaMask/metamask-extension/pull/16733#discussion_r1037342053) to make it extensible for other types of confirmations once we've decided the overall approach is sound.
- [ ] Write up ticket for further UX/technical investigations around allowing users to select between different RPC providers when prompted to switch to a given chainId (`wallet_switchEthereumChain`)
- [ ] Write up ticket further UXinvestigations around allowing users to prevent further `wallet_addEthereumChain` requests for networks they already have added - disallow dapps prompting them to add alternative RPC providers


### Before

https://user-images.githubusercontent.com/34557516/205111194-2fb4a932-656c-4cd7-8584-bc9064c979ca.mov

### After
https://user-images.githubusercontent.com/34557516/205102495-b5261ba7-7e5a-4177-92cb-9daeabe16ae6.mov

## Manual Testing Steps
1. Navigate to any site (just to make use of injected provider)
2. Open the console and paste:
```
ethereum.request({
        method: 'wallet_addEthereumChain',
        params: [
          {
            chainId: '0x1',
            chainName: 'Flashbots Protect RPC',
            rpcUrls: ['https://rpc.flashbots.net'],
		nativeCurrency:  {
		symbol: 'ETH' ,decimals: 18
          },
}
        ],
      });
```
3. Go through confirmation steps - **review extra warning modal UX** - and approve.
4. See that `Flashbots Protect RPC` (alternative Mainnet RPC) has been added as a custom network

## Pre-merge author checklist
